### PR TITLE
[19377] Add ThreadSettings to Qos

### PR DESCRIPTION
**Pull request by EduPonz**
## Description

This PR adds Qos support for configuring the various Fast DDS threads

## Contributor Checklist
- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- [x] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- [x] Any new/modified methods have been properly documented using Doxygen. 
- [ ] Changes are ABI compatible. 
- [x] Changes are API compatible. 
- [ ] New feature has been added to the `versions.md` file (if applicable).
- [ ] New feature has been documented/Current behavior is correctly described in the documentation. 
    
- *N/A*: Applicable backports have been included in the description.

## Reviewer Checklist
- [x] The PR has a milestone assigned.
- [x] Check contributor checklist is correct.
- [ ] Check CI results: changes do not issue any warning.
- [ ] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

**Comment by EduPonz** [comment]
@richiprosima please test this

File: include/fastdds/dds/domain/qos/DomainParticipantQos.hpp [hunk 1]
```diff
@@ -350,6 +447,15 @@ class DomainParticipantQos
...
     FlowControllerDescriptorList flow_controllers_;
 
+    //! Thread settings for the timed events thread
+    rtps::ThreadSettings timed_events_thread_;
+
+    //! Thread settings for the discovery server thread
+    rtps::ThreadSettings discovery_server_thread_;
+
+    //! Thread settings for the security log thread
+    rtps::ThreadSettings security_log_thread_;
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply include/fastdds/dds/domain/qos/DomainParticipantQos.hpp:491]: cf7336183

File: src/cpp/fastdds/domain/DomainParticipantImpl.cpp [hunk 2]
```diff
@@ -2320,6 +2320,27 @@ bool DomainParticipantImpl::can_qos_be_updated(
         updatable = false;
         EPROSIMA_LOG_WARNING(RTPS_QOS_CHECK, "Participant name cannot be changed after the participant is enabled");
     }
+    // TODO(eduponz): discuss possibility of updating thread settings
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply src/cpp/fastdds/domain/DomainParticipantImpl.cpp:2325]: cf7336183

File: include/fastdds/dds/domain/qos/DomainParticipantQos.hpp [hunk 3]
```diff
@@ -447,6 +478,9 @@ class DomainParticipantQos
      */
     FlowControllerDescriptorList flow_controllers_;
 
+    //! Thread settings for the builtin flow controllers sender threads
+    rtps::ThreadSettings builtin_controllers_sender_thread_;
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply include/fastdds/dds/domain/qos/DomainParticipantQos.hpp:482]: cf7336183

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp [hunk 4]
```diff
@@ -1556,6 +1556,13 @@ bool DataReaderImpl::can_qos_be_updated(
         EPROSIMA_LOG_WARNING(RTPS_QOS_CHECK,
                 "Unique network flows request cannot be changed after the creation of a DataReader.");
     }
+    // TODO(eduponz): discuss possibility of updating thread settings
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply src/cpp/fastdds/subscriber/DataReaderImpl.cpp:1559]: 170910e16

File: test/unittest/dds/participant/ParticipantTests.cpp [hunk 5]
```diff
@@ -2979,12 +2979,11 @@ TEST(ParticipantTests, SetDomainParticipantQos)
 }
 
 /*
- * This test checks that the PropertyPolicyQos and TransportConfigQos are immutable policy qos, i.e. these can not be
+ * This test checks that the mmutable policy qos can not be
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply test/unittest/dds/participant/ParticipantTests.cpp:2982]: f8a721b8b

File: src/cpp/fastdds/utils/QosConverters.cpp [hunk 6]
```diff
@@ -196,6 +196,7 @@ void set_attributes_from_qos(
     attr.listenSocketBufferSize = qos.transport().listen_socket_buffer_size;
     attr.userData = qos.user_data().data_vec();
     attr.flow_controllers = qos.flow_controllers();
+    attr.builtin_controllers_sender_thread = qos.builtin_controllers_sender_thread();
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply src/cpp/fastdds/utils/QosConverters.cpp:199]: 4f8c673b4

File: include/fastdds/rtps/transport/TransportDescriptorInterface.h [hunk 7]
```diff
@@ -39,10 +41,13 @@ class TransportInterface;
  *
  * @ingroup RTPS_MODULE
  * */
-struct RTPS_DllAPI TransportDescriptorInterface
+struct TransportDescriptorInterface
 {
+
+    using ReceptionThreadsConfigMap = std::map<uint32_t, ThreadSettings>;
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply include/fastdds/rtps/transport/TransportDescriptorInterface.h]: d0f05abc5

File: include/fastdds/rtps/transport/TransportDescriptorInterface.h [hunk 8]
```diff
@@ -51,45 +56,61 @@ struct RTPS_DllAPI TransportDescriptorInterface
...
      * (maximum number of guessed initial peers to try to connect)
      */
-    virtual uint32_t max_initial_peers_range() const
+    virtual RTPS_DllAPI uint32_t max_initial_peers_range() const
     {
         return maxInitialPeersRange;
     }
 
+    //! Returns the maximum size expected for received messages.
+    virtual RTPS_DllAPI ThreadSettings default_reception_threads() const
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply include/fastdds/rtps/transport/TransportDescriptorInterface.h]: f29daf8da

File: include/fastdds/rtps/transport/TransportDescriptorInterface.h [hunk 9]
```diff
@@ -51,45 +56,61 @@ struct RTPS_DllAPI TransportDescriptorInterface
...
+    {
+        return defaultReceptionThreads;
+    }
+
+    /** Returns the maximum number of opened channels for each initial remote peer
+     * (maximum number of guessed initial peers to try to connect)
+     */
+    virtual RTPS_DllAPI ReceptionThreadsConfigMap reception_threads() const
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply include/fastdds/rtps/transport/TransportDescriptorInterface.h]: f29daf8da

**Comment by EduPonz** [comment]
@richiprosima please test this

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: include/fastdds/dds/domain/qos/DomainParticipantQos.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply include/fastdds/dds/domain/qos/DomainParticipantQos.hpp:488]: This is pending to be added to the immutable QoS test

[the remark this answers is not in the document]
Comment (EduPonz) [reply include/fastdds/dds/domain/qos/DomainParticipantQos.hpp:488]: I think it is (see [here](https://github.com/eProsima/Fast-DDS/commit/889e062c3c1f3096b48a71d8349a2eb97f05cf8c#diff-2db41f0312d81abca828bec206b7ba1a84a3bca4afdea627022291a077cd3731R2325)).

File: test/unittest/dds/subscriber/DataReaderTests.cpp [hunk 10]
```diff
@@ -3545,6 +3545,103 @@ TEST_F(DataReaderTests, CustomPoolCreation)
...
+    reader_qos.reader_resource_limits().matched_publisher_allocation.maximum =
+            reader_qos.reader_resource_limits().matched_publisher_allocation.maximum - 1;
+    ASSERT_EQ(ReturnCode_t::RETCODE_IMMUTABLE_POLICY, data_reader->set_qos(reader_qos));
+
+    // Datasharing
+    reader_qos = DATAREADER_QOS_DEFAULT;
+    reader_qos.data_sharing().off();
+    ASSERT_EQ(ReturnCode_t::RETCODE_IMMUTABLE_POLICY, data_reader->set_qos(reader_qos));
+
+    // DataSharingQosPolicy has no API to set shm_direcory
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply test/unittest/dds/subscriber/DataReaderTests.cpp]: dfb2c3a8b

File: include/fastdds/rtps/attributes/RTPSParticipantAttributes.h [hunk 11]
```diff
@@ -576,6 +579,18 @@ class RTPSParticipantAttributes
...
+    fastdds::rtps::ThreadSettings builtin_controllers_sender_thread;
+
+    //! Thread settings for the timed events thread
+    fastdds::rtps::ThreadSettings timed_events_thread;
+
+    //! Thread settings for the discovery server thread
+    fastdds::rtps::ThreadSettings discovery_server_thread;
+
+    //! Thread settings for the security log thread
+    fastdds::rtps::ThreadSettings security_log_thread;
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply include/fastdds/rtps/attributes/RTPSParticipantAttributes.h:593]: 4e4f4fda7

File: include/fastdds/rtps/transport/PortBasedTransportDescriptor.hpp [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+
+    //! Comparison operator
+    bool RTPS_DllAPI operator ==(
+            const PortBasedTransportDescriptor& t) const;
+
+    //! Returns the ThreadSettings for the default reception threads
+    virtual RTPS_DllAPI const ThreadSettings& default_reception_threads() const;
+
+    //! Returns the ThreadSettings for the user-configured reception threads
+    virtual RTPS_DllAPI const ReceptionThreadsConfigMap& reception_threads() const;
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply include/fastdds/rtps/transport/PortBasedTransportDescriptor.hpp]: 11389c67a

File: include/fastdds/rtps/transport/PortBasedTransportDescriptor.hpp [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply include/fastdds/rtps/transport/PortBasedTransportDescriptor.hpp:90]: 11389c67a

File: include/fastdds/rtps/transport/PortBasedTransportDescriptor.hpp [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+/**
+ * Base class for all port based transport descriptors
+ *
+ * This class provides a common thread settings configuration for all
+ * port based transport descriptor implementations
+ *
+ * @ingroup TRANSPORT_MODULE
+ */
+class PortBasedTransportDescriptor : public TransportDescriptorInterface
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply include/fastdds/rtps/transport/PortBasedTransportDescriptor.hpp:37]: 11389c67a

File: include/fastdds/rtps/transport/TransportDescriptorInterface.h [hunk 15]
```diff
@@ -15,12 +15,13 @@
 #ifndef _FASTDDS_TRANSPORT_DESCRIPTOR_INTERFACE_H_
 #define _FASTDDS_TRANSPORT_DESCRIPTOR_INTERFACE_H_
 
-#include <fastrtps/fastrtps_dll.h>
-
 #include <cstdint>
+#include <map>
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply include/fastdds/rtps/transport/TransportDescriptorInterface.h]: 11389c67a

File: include/fastdds/rtps/transport/PortBasedTransportDescriptor.hpp [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+    //! Set the ThreadSettings for the default reception threads
+    virtual RTPS_DllAPI void default_reception_threads(
+            const ThreadSettings& default_reception_threads);
+
+    //! Returns the ThreadSettings for the user-configured reception threads
+    virtual RTPS_DllAPI const ReceptionThreadsConfigMap& reception_threads() const;
+
+    //! Set the ThreadSettings for the user-configured reception threads
+    virtual RTPS_DllAPI void reception_threads(
+            const ReceptionThreadsConfigMap& reception_threads);
```
Comment (MiguelCompany) [inline include/fastdds/rtps/transport/PortBasedTransportDescriptor.hpp:94]: Do we need all these methods to be virtual ?

File: include/fastdds/rtps/transport/PortBasedTransportDescriptor.hpp [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+
```
Comment (MiguelCompany) [inline include/fastdds/rtps/transport/PortBasedTransportDescriptor.hpp:95]: Also missing the other "setter" methods:

`ReceptionThreadsConfigMap& reception_threads();`

File: test/unittest/transport/CMakeLists.txt [hunk 18]
```diff
@@ -269,9 +284,44 @@ endif()
...
+    $<$<BOOL:${ANDROID}>:${ANDROID_IFADDRS_INCLUDE_DIR}>
+    )
+target_link_libraries(test_UDPv4Tests GTest::gtest ${MOCKS})
+if(QNX)
+    target_link_libraries(test_UDPv4Tests socket)
+endif()
+if(MSVC OR MSVC_IDE)
+    target_link_libraries(test_UDPv4Tests ${PRIVACY} iphlpapi Shlwapi)
+else()
+    target_link_libraries(test_UDPv4Tests ${PRIVACY})
```
Comment (MiguelCompany) [inline test/unittest/transport/CMakeLists.txt:311]:
```suggestion
```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: include/fastdds/rtps/transport/PortBasedTransportDescriptor.hpp (hunk 16 above)
Comment (EduPonz) [reply include/fastdds/rtps/transport/PortBasedTransportDescriptor.hpp]: 43be267bb

File: include/fastdds/rtps/transport/PortBasedTransportDescriptor.hpp (hunk 17 above)
Comment (EduPonz) [reply include/fastdds/rtps/transport/PortBasedTransportDescriptor.hpp:99]: 43be267bb

File: test/unittest/transport/CMakeLists.txt (hunk 18 above)
Comment (EduPonz) [reply test/unittest/transport/CMakeLists.txt]: 43be267bb

File: include/fastdds/dds/domain/qos/DomainParticipantQos.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (EduPonz) [reply include/fastdds/dds/domain/qos/DomainParticipantQos.hpp:488]: 354fb2ef4

**Comment by EduPonz** [comment]
@richiprosima please test this

**Review by MiguelCompany — APPROVED** [review]

**Review by MiguelCompany — APPROVED** [review]
LGTM

**Comment by MiguelCompany** [comment]
@richiprosima Please test this

File: include/fastdds/dds/subscriber/qos/DataReaderQos.hpp [hunk 19]
```diff
@@ -920,6 +951,9 @@ class DataReaderQos
 
     //!DataSharing configuration (Extension)
     DataSharingQosPolicy data_sharing_;
+
+    //! Thread settings for the data-sharing listener thread
+    rtps::ThreadSettings data_sharing_listener_thread_;
```
Comment (MiguelCompany) [inline include/fastdds/dds/subscriber/qos/DataReaderQos.hpp:956]: This needs to be propagated into `rtps::ReaderAttributes`

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: include/fastdds/dds/subscriber/qos/DataReaderQos.hpp (hunk 19 above)
Comment (EduPonz) [reply include/fastdds/dds/subscriber/qos/DataReaderQos.hpp:956]: d278ba0fb

**Comment by EduPonz** [comment]
@richiprosima please test mac

**Comment by EduPonz** [comment]
@richiprosima please test this

**Pull request merged**